File: python/src/rules.py
"""Evaluate per-product notification rules."""
from __future__ import annotations

from decimal import Decimal
from typing import Any, Dict, List, Optional, Tuple


def _fmt_money(value: Decimal, currency: str) -> str:
    symbol = {"GBP": "£", "USD": "$", "EUR": "€"}.get(currency, "")
    return f"{symbol}{value:.2f}"


def _threshold_key(rule: Dict[str, Any]) -> str:
    return f"threshold:{Decimal(str(rule['at_or_below'])):.2f}"


def evaluate(
    rules: List[Dict[str, Any]],
    old_price: Optional[Decimal],
    new_price: Decimal,
    currency: str,
    armed_rules: Optional[Dict[str, bool]] = None,
) -> Tuple[List[str], Dict[str, bool]]:
    """Evaluate `rules` and return (reasons, updated_armed_state).

    `armed_rules` is the previous arming state (dict of rule key -> armed bool).
    A rule key is treated as armed by default (missing == True).
    """
    armed_state = dict(armed_rules or {})
    reasons: List[str] = []

    for rule in rules:
        t = rule.get("type")

        if t == "any_change":
            if old_price is not None and new_price != old_price:
                reasons.append("price changed")

        elif t == "threshold":
            at_or_below = Decimal(str(rule["at_or_below"]))
            only_once = bool(rule.get("only_once", False))
            key = _threshold_key(rule)
            currently_below = new_price <= at_or_below
            previously_below = old_price is not None and old_price <= at_or_below

            if only_once:
                # Re-arm when price moves above the threshold
                if not currently_below:
                    armed_state[key] = True
                else:
                    is_armed = armed_state.get(key, True)
                    if is_armed and not previously_below:
                        reasons.append(f"hit target {_fmt_money(at_or_below, currency)}")
                        armed_state[key] = False
                    elif is_armed and old_price is None:
                        # First time seeing this product and it's already at/below target
                        reasons.append(f"at target {_fmt_money(at_or_below, currency)}")
                        armed_state[key] = False
            else:
                if currently_below and not previously_below:
                    reasons.append(f"hit target {_fmt_money(at_or_below, currency)}")
                elif currently_below and old_price is None:
                    reasons.append(f"at target {_fmt_money(at_or_below, currency)}")

        elif t == "percent_drop":
            min_percent = Decimal(str(rule["min_percent"]))
            if old_price is not None and old_price > 0 and new_price < old_price:
                drop_pct = (old_price - new_price) / old_price * Decimal(100)
                if drop_pct >= min_percent:
                    reasons.append(f"-{drop_pct:.1f}% drop")

    # De-duplicate while preserving order
    seen = set()
    deduped = []
    for r in reasons:
        if r not in seen:
            deduped.append(r)
            seen.add(r)
    return deduped, armed_state

File: rust/src/rules.rs
use crate::config::Rule;
use rust_decimal::Decimal;
use rust_decimal_macros::dec;
use std::collections::BTreeMap;

fn currency_symbol(c: &str) -> &str {
    match c {
        "GBP" => "£",
        "USD" => "$",
        "EUR" => "€",
        _ => "",
    }
}

fn fmt_money(value: Decimal, currency: &str) -> String {
    format!("{}{:.2}", currency_symbol(currency), value)
}

fn threshold_key(at_or_below: Decimal) -> String {
    format!("threshold:{:.2}", at_or_below)
}

pub fn evaluate(
    rules: &[Rule],
    old_price: Option<Decimal>,
    new_price: Decimal,
    currency: &str,
    armed_prev: &BTreeMap<String, bool>,
) -> (Vec<String>, BTreeMap<String, bool>) {
    let mut armed = armed_prev.clone();
    let mut reasons: Vec<String> = Vec::new();

    for rule in rules {
        match rule {
            Rule::AnyChange => {
                if let Some(old) = old_price {
                    if new_price != old {
                        reasons.push("price changed".into());
                    }
                }
            }
            Rule::Threshold { at_or_below, only_once } => {
                let target = Decimal::try_from(*at_or_below).unwrap_or(dec!(0));
                let key = threshold_key(target);
                let currently_below = new_price <= target;
                let previously_below = matches!(old_price, Some(p) if p <= target);

                if *only_once {
                    if !currently_below {
                        armed.insert(key.clone(), true);
                    } else {
                        let is_armed = armed.get(&key).copied().unwrap_or(true);
                        if is_armed && !previously_below {
                            reasons.push(format!("hit target {}", fmt_money(target, currency)));
                            armed.insert(key, false);
                        } else if is_armed && old_price.is_none() {
                            reasons.push(format!("at target {}", fmt_money(target, currency)));
                            armed.insert(key, false);
                        }
                    }
                } else if currently_below && !previously_below {
                    reasons.push(format!("hit target {}", fmt_money(target, currency)));
                } else if currently_below && old_price.is_none() {
                    reasons.push(format!("at target {}", fmt_money(target, currency)));
                }
            }
            Rule::PercentDrop { min_percent } => {
                let min = Decimal::try_from(*min_percent).unwrap_or(dec!(0));
                if let Some(old) = old_price {
                    if old > dec!(0) && new_price < old {
                        let drop_pct = (old - new_price) / old * dec!(100);
                        if drop_pct >= min {
                            reasons.push(format!("-{:.1}% drop", drop_pct));
                        }
                    }
                }
            }
        }
    }

    // Dedupe preserving order.
    let mut seen = std::collections::HashSet::new();
    let deduped: Vec<String> = reasons
        .into_iter()
        .filter(|r| seen.insert(r.clone()))
        .collect();
    (deduped, armed)
}
